File: Client/jSpamBot/src/main/java/IPBan.java
/*
 * Copyright (c) 2005-2018, BearWare.dk
 *
 * Contact Information:
 *
 * Bjoern D. Rasmussen
 * Kirketoften 5
 * DK-8260 Viby J
 * Denmark
 * Email: contact@bearware.dk
 * Phone: +45 20 20 54 59
 * Web: http://www.bearware.dk
 *
 * This source code is part of the TeamTalk SDK owned by
 * BearWare.dk. Use of this file, or its compiled unit, requires a
 * TeamTalk SDK License Key issued by BearWare.dk.
 *
 * The TeamTalk SDK License Agreement along with its Terms and
 * Conditions are outlined in the file License.txt included with the
 * TeamTalk SDK distribution.
 *
 */

import dk.bearware.BanType;
import dk.bearware.BannedUser;
import dk.bearware.TeamTalkBase;
import dk.bearware.UserAccount;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.time.Duration;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Vector;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class IPBan {
    TimeProvider time;
    // bans already on server (doListBans() result)
    Vector<BannedUser> serverbans = new Vector<>();
    // IPs/networks to ban (loaded from file)
    Vector<String> networks;
    int networkindex = -1;
    Duration banDuration;
    Logger logger;
    // Ban types made by spambot (ban network and ban user)
    final int SPAMBOT_NETWORK_BAN_TYPE = BanType.BANTYPE_IPADDR;
    final int SPAMBOT_USER_BAN_TYPE = BanType.BANTYPE_IPADDR | BanType.BANTYPE_USERNAME;

    public IPBan(TimeProvider time, Vector<String> networks, Duration banDuration, Logger log) {
        this.time = time;
        this.networks = networks;
        this.banDuration = banDuration;
        this.logger = log;
    }

    static void setBanTime(BannedUser ban) {
        ZonedDateTime now = ZonedDateTime.now(ZoneId.of("UTC"));
        ban.szBanTime = String.format("%04d/%02d/%02d %02d:%02d",
                now.getYear(),
                now.getMonthValue(),
                now.getDayOfMonth(),
                now.getHour(),
                now.getMinute());
    }

    public static Vector<String> loadFile(String filename, Logger log) {
        Vector<String> networks = new Vector<>();
        File file = new File(filename);
        if (file.exists()) {
            try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                String line;
                while ((line = br.readLine()) != null) {
                    networks.add(line);
                }
            }
            catch(IOException e) {
                log.severe("Failed to load " + filename + ": " + e);
            }
        }
        return networks;
    }

    private void addBan(BannedUser ban) {
        if (ban.szBanTime.isEmpty()) {
            logger.severe("Ban time is not set on ban owner " + ban.szOwner + " IP addr: " + ban.szIPAddress);
        }
        if (ban.szOwner.isEmpty()) {
            logger.severe("Ban owner is not set on ban type " + ban.uBanTypes + " IP addr: " + ban.szIPAddress);
        }
        if (ban.uBanTypes == BanType.BANTYPE_NONE) {
            logger.severe("Ban type is not set, IP addr: " + ban.szIPAddress);
        }
        serverbans.add(ban);
    }

    public void addLocalBan(BannedUser ban, TeamTalkBase ttinst) {
        UserAccount ua = new UserAccount();
        ttinst.getMyUserAccount(ua);
        ban.szOwner = ua.szUsername;
        setBanTime(ban);
        addRemoteBan(ban);
    }

    public void addRemoteBan(BannedUser ban) {
        addBan(ban);
    }

    private void removeBan(BannedUser ban) {
        this.serverbans.remove(ban);
    }

    public int syncBans(TeamTalkBase ttinst) {
        // WebLogin name (spambot@bearware.dk)
        UserAccount ua = new UserAccount();
        ttinst.getMyUserAccount(ua);

        // make list of network bans created by spambot
        Vector<BannedUser> mybans = new Vector<>();
        Vector<String> ipaddrs = new Vector<>();
        for (BannedUser b : this.serverbans) {
            if (b.uBanTypes == SPAMBOT_NETWORK_BAN_TYPE && b.szOwner.equals(ua.szUsername)) {
                mybans.add(b);
                ipaddrs.add(b.szIPAddress);
            }
        }

        //remove network bans no longer found in 'networks'
        for (BannedUser b : mybans) {
            if (!this.networks.contains(b.szIPAddress)) {
                this.serverbans.remove(b);
                this.logger.info(String.format("Removed missing network ban: 0x%x IP: %s Username: %s Channel: %s",
                                               b.uBanTypes, b.szIPAddress, b.szUsername, b.szChannelPath));
                return ttinst.doUnBanUserEx(b);
            }
        }

        // submit new network bans not already on server
        for (++this.networkindex; this.networkindex < networks.size();++this.networkindex) {
            String ipaddr = networks.elementAt(this.networkindex);
            if (!ipaddrs.contains(ipaddr)) {
                BannedUser b = new BannedUser();
                b.szIPAddress = ipaddr;
                b.uBanTypes = SPAMBOT_NETWORK_BAN_TYPE;
                this.logger.info(String.format("Added new network ban: 0x%x IP: %s Username: %s Channel: %s",
                                               b.uBanTypes, b.szIPAddress, b.szUsername, b.szChannelPath));
                // store in local ban list
                int cmdid = ttinst.doBan(b);
                if (cmdid > 0) {
                    addLocalBan(b, ttinst);
                }
                return cmdid;
            }
        }

        // remove expired user bans
        if (!this.banDuration.isZero()) {
            ZonedDateTime bannedBefore = ZonedDateTime.now(ZoneId.of("UTC")).minus(this.banDuration);
            for (BannedUser b : getBannedBySpamBot(ua, SPAMBOT_USER_BAN_TYPE, bannedBefore)) {
                removeBan(b);
                this.logger.info(String.format("Removed expired ban: 0x%x IP: %s Username: %s Channel: %s",
                                               b.uBanTypes, b.szIPAddress, b.szUsername, b.szChannelPath));
                return ttinst.doUnBanUserEx(b);
            }
        }

        return 0;
    }

    private Vector<BannedUser> getBannedBySpamBot(UserAccount ua, int uBanTypes, ZonedDateTime before) {
        Vector<BannedUser> result = new Vector<>();
        for (BannedUser b : getBannedUsersBefore(before)) {
            if (b.szOwner.equals(ua.szUsername) && b.uBanTypes == uBanTypes) {
                result.add(b);
            }
        }
        return result;
    }

    private Vector<BannedUser> getBannedUsersBefore(ZonedDateTime zdt) {
        Vector<BannedUser> result = new Vector<>();
        for (BannedUser b : this.serverbans) {
            Pattern pattern = Pattern.compile("^(\\d+)/(\\d+)/(\\d+) (\\d+):(\\d+)$");
            Matcher datematch = pattern.matcher(b.szBanTime);
            if (datematch.find()) {
                ZonedDateTime bantime = ZonedDateTime.of(Integer.parseInt(datematch.group(1)), //year
                                                         Integer.parseInt(datematch.group(2)), //month
                                                         Integer.parseInt(datematch.group(3)), //day
                                                         Integer.parseInt(datematch.group(4)), //hour
                                                         Integer.parseInt(datematch.group(5)), //minute
                                                         0, 0, //seconds, nanos
                                                         ZoneId.of("UTC"));
                if (bantime.isBefore(zdt)) {
                    result.add(b);
                }
            }
            else {
                this.logger.info(String.format("Date format %s does not match: ", b.szBanTime));
            }
        }
        return result;
    }
}

File: Library/TeamTalkLib/bin/ttspambot/IPBan.cpp
/*
 * Copyright (c) 2005-2018, BearWare.dk
 *
 * Contact Information:
 *
 * Bjoern D. Rasmussen
 * Kirketoften 5
 * DK-8260 Viby J
 * Denmark
 * Email: contact@bearware.dk
 * Phone: +45 20 20 54 59
 * Web: http://www.bearware.dk
 *
 * This source code is part of the TeamTalk SDK owned by
 * BearWare.dk. Use of this file, or its compiled unit, requires a
 * TeamTalk SDK License Key issued by BearWare.dk.
 *
 * The TeamTalk SDK License Agreement along with its Terms and
 * Conditions are outlined in the file License.txt included with the
 * TeamTalk SDK distribution.
 *
 */

#include "IPBan.h"

#include "teamtalk/Log.h"

#include <ace/Log_Msg.h>
#include <ace/OS_NS_stdio.h>
#include <ace/OS_NS_string.h>

#include <algorithm>
#include <chrono>
#include <ctime>
#include <fstream>
#include <regex>
#include <string>

namespace spambot {

    namespace {

        std::string Tstr(const TTCHAR* p)
        {
#if defined(UNICODE)
            std::wstring ws(p);
            return std::string(ws.begin(), ws.end());
#else
            return std::string(p);
#endif
        }

        void SetTstr(TTCHAR* dst, std::size_t cap, const std::string& src)
        {
#if defined(UNICODE)
            std::wstring ws(src.begin(), src.end());
            std::size_t n = std::min(ws.size(), cap - 1);
            std::copy_n(ws.begin(), n, dst);
            dst[n] = 0;
#else
            std::size_t n = std::min(src.size(), cap - 1);
            std::copy_n(src.begin(), n, dst);
            dst[n] = 0;
#endif
        }

    } // anonymous namespace

    IPBan::IPBan(std::vector<std::string> networks, Duration banDuration)
        : m_networks(std::move(networks)), m_banDuration(banDuration)
    {
    }

    std::vector<std::string> IPBan::LoadFile(const std::string& filename)
    {
        std::vector<std::string> networks;
        if (filename.empty())
            return networks;

        std::ifstream f(filename);
        if (!f.is_open())
        {
            ACE_TCHAR msg[512];
            ACE_OS::snprintf(msg, 512, ACE_TEXT("Failed to load VPN IP file %C."), filename.c_str());
            TT_SYSLOG(msg);
            return networks;
        }

        std::string line;
        while (std::getline(f, line))
        {
            if (!line.empty() && line.back() == '\r')
                line.pop_back();
            if (!line.empty())
                networks.push_back(line);
        }
        return networks;
    }

    void IPBan::SetBanTimeUTC(BannedUser& ban)
    {
        auto now = std::chrono::system_clock::now();
        std::time_t t = std::chrono::system_clock::to_time_t(now);
        std::tm tm_utc{};
#if defined(WIN32)
        gmtime_s(&tm_utc, &t);
#else
        gmtime_r(&t, &tm_utc);
#endif
        char buf[32];
        std::snprintf(buf, sizeof(buf), "%04d/%02d/%02d %02d:%02d",
                      tm_utc.tm_year + 1900, tm_utc.tm_mon + 1, tm_utc.tm_mday,
                      tm_utc.tm_hour, tm_utc.tm_min);
        SetTstr(ban.szBanTime, TT_STRLEN, buf);
    }

    void IPBan::AddLocalBan(BannedUser& ban, TTInstance* ttinst)
    {
        UserAccount ua{};
        TT_GetMyUserAccount(ttinst, &ua);
        SetTstr(ban.szOwner, TT_STRLEN, Tstr(ua.szUsername));
        SetBanTimeUTC(ban);
        AddRemoteBan(ban);
    }

    void IPBan::AddRemoteBan(const BannedUser& ban)
    {
        ACE_TCHAR msg[512];
        if (Tstr(ban.szBanTime).empty())
        {
            ACE_OS::snprintf(msg, 512, ACE_TEXT("Ban time is not set, IP: %C."), Tstr(ban.szIPAddress).c_str());
            TT_SYSLOG(msg);
        }
        if (Tstr(ban.szOwner).empty())
        {
            ACE_OS::snprintf(msg, 512, ACE_TEXT("Ban owner is not set, IP: %C."), Tstr(ban.szIPAddress).c_str());
            TT_SYSLOG(msg);
        }
        if (ban.uBanTypes == BANTYPE_NONE)
        {
            ACE_OS::snprintf(msg, 512, ACE_TEXT("Ban type is not set, IP: %C."), Tstr(ban.szIPAddress).c_str());
            TT_SYSLOG(msg);
        }
        m_serverbans.push_back(ban);
    }

    void IPBan::ClearServerBans()
    {
        m_serverbans.clear();
        m_networkindex = -1;
    }

    int IPBan::SyncBans(TTInstance* ttinst)
    {
        UserAccount ua{};
        TT_GetMyUserAccount(ttinst, &ua);
        const std::string myname = Tstr(ua.szUsername);

        std::vector<BannedUser> mybans;
        std::vector<std::string> ipaddrs;
        for (const auto& b : m_serverbans)
        {
            if (b.uBanTypes == NETWORK_BAN_TYPE && Tstr(b.szOwner) == myname)
            {
                mybans.push_back(b);
                ipaddrs.push_back(Tstr(b.szIPAddress));
            }
        }

        for (const auto& b : mybans)
        {
            const std::string ip = Tstr(b.szIPAddress);
            if (std::find(m_networks.begin(), m_networks.end(), ip) == m_networks.end())
            {
                m_serverbans.erase(std::remove_if(m_serverbans.begin(), m_serverbans.end(),
                    [&](const BannedUser& x){ return Tstr(x.szIPAddress) == ip
                                                  && x.uBanTypes == b.uBanTypes
                                                  && Tstr(x.szOwner) == myname; }),
                    m_serverbans.end());
                ACE_TCHAR msg[512];
                ACE_OS::snprintf(msg, 512, ACE_TEXT("Removed missing network ban: IP %C."), ip.c_str());
                TT_LOG(msg);
                return TT_DoUnBanUserEx(ttinst, &b);
            }
        }

        for (++m_networkindex; m_networkindex < static_cast<int>(m_networks.size()); ++m_networkindex)
        {
            const std::string& ip = m_networks[m_networkindex];
            if (std::find(ipaddrs.begin(), ipaddrs.end(), ip) == ipaddrs.end())
            {
                BannedUser nb{};
                SetTstr(nb.szIPAddress, TT_STRLEN, ip);
                nb.uBanTypes = NETWORK_BAN_TYPE;
                ACE_TCHAR msg[512];
                ACE_OS::snprintf(msg, 512, ACE_TEXT("Added new network ban: IP %C."), ip.c_str());
                TT_LOG(msg);
                int cmdid = TT_DoBan(ttinst, &nb);
                if (cmdid > 0)
                    AddLocalBan(nb, ttinst);
                return cmdid;
            }
        }

        if (m_banDuration.count() > 0)
        {
            auto cutoff = std::chrono::system_clock::now() - m_banDuration;
            std::time_t cutoff_t = std::chrono::system_clock::to_time_t(cutoff);
            std::tm cutoff_tm{};
#if defined(WIN32)
            gmtime_s(&cutoff_tm, &cutoff_t);
#else
            gmtime_r(&cutoff_t, &cutoff_tm);
#endif
            std::regex re(R"(^(\d+)/(\d+)/(\d+) (\d+):(\d+)$)");
            for (auto it = m_serverbans.begin(); it != m_serverbans.end(); ++it)
            {
                if (it->uBanTypes != USER_BAN_TYPE || Tstr(it->szOwner) != myname)
                    continue;
                std::smatch m;
                std::string bantime = Tstr(it->szBanTime);
                if (!std::regex_search(bantime, m, re))
                    continue;
                std::tm bt{};
                bt.tm_year = std::stoi(m[1]) - 1900;
                bt.tm_mon  = std::stoi(m[2]) - 1;
                bt.tm_mday = std::stoi(m[3]);
                bt.tm_hour = std::stoi(m[4]);
                bt.tm_min  = std::stoi(m[5]);

                auto cmp = [](const std::tm& a, const std::tm& b) {
                    if (a.tm_year != b.tm_year) return a.tm_year < b.tm_year;
                    if (a.tm_mon  != b.tm_mon ) return a.tm_mon  < b.tm_mon;
                    if (a.tm_mday != b.tm_mday) return a.tm_mday < b.tm_mday;
                    if (a.tm_hour != b.tm_hour) return a.tm_hour < b.tm_hour;
                    return a.tm_min < b.tm_min;
                };
                if (cmp(bt, cutoff_tm))
                {
                    BannedUser b = *it;
                    m_serverbans.erase(it);
                    ACE_TCHAR msg[512];
                    ACE_OS::snprintf(msg, 512, ACE_TEXT("Removed expired ban: IP %C."), Tstr(b.szIPAddress).c_str());
                    TT_LOG(msg);
                    return TT_DoUnBanUserEx(ttinst, &b);
                }
            }
        }

        return 0;
    }

} // namespace spambot
